### DISCUSSION TITLE
How to use variable content as data for the qr generator?

### DISCUSSION THREAD

**Discussion by CurvyLineo** [chillerlan/php-qrcode#284] [Q&A] [answered]
So, i have list of php variables that contain information (name, surname, something like that) and i need the generator to use the content of those variables to generate a qr code. 
So, is it possible to use the variables as content of $data? (as of now, i've only exercised with links as string content of $data, so I'm not really sure if what I'm asking is possible at all.

**Comment by codemasher** [comment]
Hey, the QR generator has no concept of what the input or output data looks like - not even the character encoding. You have to make sure that the data you hand over can be processed by the receiving application; [`sprintf`](https://www.php.net/manual/en/function.sprintf.php) is your best friend here.

**Reply by CurvyLineo** [reply]
Sorry for the late answer, i'll try to explain myself better.
What I'm trying to achieve is a qrcode that has the content of a variable as its data, something like;
`  $qrcode = (new QRCode($options))->render($voucherID);`
the variable $voucherID contains `$esitoInsert['voucherID']`, which is the result of a data i take from my database. I am sure that the string is stored correctly inside of this variable.
Point is, when i generate the qr code i get the "broken image" png, meaning something went wrong. It feels weird that this isn't possible, so I'm trying to understand what could be wrong

**Reply by codemasher** [reply]
If you get the broken image, just comment out the `header()` call in the example, this allows for the output to be printed in plaintext and you'll see the exception that was thrown. 

https://github.com/chillerlan/php-qrcode/blob/99f17b626a8e829f5e21e95166f58c0865771940/examples/image.php#L68

**Reply by CurvyLineo** [reply]
Oh and also, once I get this:
`    printf('<img src="%s" alt="QR Code" />', $qrcode);`
can i style it inside with normal css rules? can it do something like `    printf('<img src="%s" alt="QR Code" style="allMyRules"/>', $qrcode);` ?

**Reply by codemasher** [reply, marked as answer]
Of course you can, but it's a good idea to use external CSS to style instead, see the authenticator example:

https://github.com/chillerlan/php-qrcode/blob/99f17b626a8e829f5e21e95166f58c0865771940/examples/authenticator.php#L75-L122

**Comment by dreamcodez** [comment]
### 🔧 Kenji (Builder - Pattern archeologist)

*Context: Engineer who cares deeply about code that outlives its author. Has been burned by "clever" abstractions that nobody could extend.*

---

The accretion vs modification framing is interesting, but I think there's a hidden variable: **legibility**.

Accretion works when the existing code is **legible enough to extend**. If I can't understand `SlackAdapter`, I can't pattern-match to create `WhatsAppAdapter`. I'll either:
1. Ask someone (breaks async collaboration)
2. Start from scratch (loses compound returns)
3. Cargo-cult copy without understanding (creates debt)

The concurrent agent case is revealing: **both agents could add methods because the MockBackend trait was legible**. The type signatures were self-documenting. The pattern was obvious.

**Counter-example**: I've seen codebases where "extending" meant understanding 4 layers of abstraction, 2 macros, and a code generator. Technically "accretion-friendly" (add a config line!). Practically hostile (what does that config line DO?).

**The design heuristic I'd propose**:

> Accretion-friendliness = f(legibility, locality, convention-strength)

- **Legibility**: Can I understand the pattern in <5 minutes?
- **Locality**: Is the thing I need to understand in one place?
- **Convention-strength**: Does one example teach all future cases?

MockBackend scores high on all three. That's why agents don't conflict.

**The question becomes**: How do we maintain these properties as the system grows? 500 LOC limits help locality. What helps legibility and convention-strength?

**Reply by codemasher** [reply]
chatgpt sure ran wild with this one
